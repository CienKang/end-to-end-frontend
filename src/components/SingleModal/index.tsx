import React, { useState } from 'react';
import { CREATE_NEW_CONTENT_TYPE, CREATE_NEW_FIELD_FOR_CONTENT_TYPE, RENAME_CONTENT_TYPE } from '../../constants/apiEndpoints';
import { ContentType, SingleModalProps } from '../../types';
import { makeRequestToBackend } from '../../utils/makeRequest/makeRequest';
import './SingleModal.css';

const SingleModal = (props: SingleModalProps) => {

    const [input, setInput] = useState<string>('');

    const handleInputChange = (e: React.ChangeEvent<HTMLInputElement>) => {
        setInput(e.target.value);
        console.log(input);
    };


    const handleCancel = () => {
        props.setShowModal(!props.showModal);
    };

    const handleCreateClickType = () => {
        makeRequestToBackend({ ...CREATE_NEW_CONTENT_TYPE }, { data: { typeName: input } }).then(() => {
            props.setShowModal(!props.showModal);
            window.location.reload();
        }
        );
    };

    const handleCreateClickField = () => {
        makeRequestToBackend({ ...CREATE_NEW_FIELD_FOR_CONTENT_TYPE(props.contentTypeSelected2.typeName) }, { data: { field: input } }).then(() => {
            props.setShowModal(!props.showModal);
            props.setContentTypeSelected2({
                ...props.contentTypeSelected2,
                fields: [...props.contentTypeSelected2.fields, input]
            });
        });
    };

    const handleCreateClickRename = () => {
        makeRequestToBackend({...RENAME_CONTENT_TYPE(props.contentTypeSelected2.typeName)},{data: {newTypeName: input}}).then(() => {
            props.setShowModal(!props.showModal);
            
            props.setContentTypes(props.contentTypes.map((contentType:ContentType) => {
                if(contentType.typeName == props.contentTypeSelected2.typeName){
                    return {
                        ...contentType,
                        typeName: input
                    };
                }
                return contentType;
            }));

            
            props.setContentTypeSelected2({
                ...props.contentTypeSelected2,
                typeName: input
            });
        });
              
    };
    return (
        <div className='single-modal'>
            <div className='single-modal-content'>
                <div className='single-modal-header'>
                    { props.modalFor !='Rename' && <h3>Create a new {props.modalFor}</h3> }
                    { props.modalFor =='Rename' && <h3>Rename {props.contentTypeSelected2.typeName}</h3> }
                </div>
                <div className='single-modal-body'>
                    <div className='combined-input'>
                        {props.modalFor !='Rename' &&  <span>Name of {props.modalFor.toLowerCase()}</span>}
                        {props.modalFor =='Rename' &&  <span>New name of {props.contentTypeSelected2.typeName}</span>}
                        <input type='text' data-testid='input-single-modal' onChange={handleInputChange} />
                    </div>

                    <div className='combined-btns'>
                        <span onClick={handleCancel}>Cancel</span>
                        {props.modalFor == 'Content Type' && <button onClick={handleCreateClickType}>Create</button>}
                        {props.modalFor == 'Field' && <button onClick={handleCreateClickField}>Create</button>}
                        {props.modalFor == 'Rename' && <button onClick={handleCreateClickRename}>Rename</button>}
                    </div>
                </div>
            </div>
        </div>
    );
};

export default SingleModal;