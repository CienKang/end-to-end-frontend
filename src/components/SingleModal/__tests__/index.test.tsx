import SingleModal from '..';
import { render, fireEvent } from '@testing-library/react';
import React from 'react';

const props = {
    modalFor: 'Content Type',
    setShowModal: jest.fn(),
    showModal: true,
    contentTypeSelected2: {},
    setContentTypeSelected2: jest.fn(),
    contentTypes: [],
    setContentTypes: jest.fn(),
};


describe('SingleModal', () => {
    it('renders without crashing', () => {
        render(<SingleModal
            modalFor='Content Type'
            setShowModal={jest.fn()}
            showModal={true}
            contentTypeSelected2={{ typeName: 'test', fields: [], id: '1' }}
            setContentTypeSelected2={jest.fn()}
            contentTypes={[]}
            setContentTypes={jest.fn()}
        />);
    });

    it('should match with previous snapshot', () => {
        const { asFragment } = render(<SingleModal
            modalFor='Content Type'
            setShowModal={jest.fn()}
            showModal={true}
            contentTypeSelected2={{ typeName: 'test', fields: [], id: '1' }}
            setContentTypeSelected2={jest.fn()}
            contentTypes={[]}
            setContentTypes={jest.fn()}
        />);
        expect(asFragment()).toMatchSnapshot();
    });

    it('should change the input value when new input is being typed', () => {
        const { getByTestId } = render(<SingleModal
            modalFor='Content Type'
            setShowModal={jest.fn()}
            showModal={true}
            contentTypeSelected2={{ typeName: 'test', fields: [], id: '1' }}
            setContentTypeSelected2={jest.fn()}
            contentTypes={[]}
            setContentTypes={jest.fn()}
        />);
        const input = getByTestId('input-single-modal');
        expect(input).toHaveValue('');
        fireEvent.change(input, { target: { value: 'test' } });
        expect(input).toHaveValue('test');
    });

    it('should call setShowModal and close single modal when cancel button is clicked', () => {

        const screen = render(<SingleModal
            modalFor='Content Type'
            setShowModal={jest.fn()}
            showModal={true}
            contentTypeSelected2={{ typeName: 'test', fields: [], id: '1' }}
            setContentTypeSelected2={jest.fn()}
            contentTypes={[]}
            setContentTypes={jest.fn()}
        />);
        const cancelButton = screen.getByText(/cancel/i);
        fireEvent.click(cancelButton);
        expect(props.setShowModal).toHaveBeenCalled();

    });

});