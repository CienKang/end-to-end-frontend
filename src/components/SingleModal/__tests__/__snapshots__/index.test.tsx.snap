// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SingleModal should match with previous snapshot 1`] = `
<DocumentFragment>
  <div
    class="single-modal"
  >
    <div
      class="single-modal-content"
    >
      <div
        class="single-modal-header"
      >
        <h3>
          Create a new Content Type
        </h3>
      </div>
      <div
        class="single-modal-body"
      >
        <div
          class="combined-input"
        >
          <span>
            Name of content type
          </span>
          <input
            data-testid="input-single-modal"
            type="text"
          />
        </div>
        <div
          class="combined-btns"
        >
          <span>
            Cancel
          </span>
          <button>
            Create
          </button>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
